### PR TITLE
Fixed the issue with parsing universal locations into the reachability query

### PR DESCRIPTION
**Pull request by Puvikaran2001**
This PR solves parsing of universal and inconsistent locations.
The current implementation in main can make the internal representation for any single location from a query.
However, when combining components, it only support locations that are not universal or inconsistent.

We have add a extra field `add` in the enum type `LocationType` to represent any location in for the particular component.

We have refactored `get_state` function and spilled some of it content into some sub functions.

The method `compare_partial_locations` in `LocationID` is removed and moved to `LocationTuple` with extra addition code to handle comparing of universal or inconsistent locations

The method `get_location` in `TransitionSystem` returns the `LocationTuple` for the given `LocationID` input argument

The method 'get_combined_decls' combines the individual declaration of components into one declaration. This method is used for `apply_constraints_to_state` in extract_state.rs

### CODE REVIEW

**Review by S3JER — DISMISSED** [review]

File: src/System/extract_state.rs [hunk 1]
```diff
@@ -21,104 +19,85 @@ pub fn get_state(
...
-                OwnedFederation::universe(system.get_dim())
-            };
+        let zone = create_zone_given_constraints(clock.as_deref(), &declarations, system)?;
 
-            Ok(State::create(locationtuple, zone))
-        }
-        _ => panic!("Expected QueryExpression::State, but got {:?}", state_query),
+        Ok(State::create(locationtuple, zone))
+    } else {
+        panic!("Expected QueryExpression::State, but got {:?}", state_query)
```
Comment (henneboy) [inline src/System/extract_state.rs:40]: maybe use unreachable! macro

**Review by henneboy — CHANGES_REQUESTED** [review]

**Review by WassawRoki — DISMISSED** [review]
LGTM

File: src/TransitionSystems/common.rs [hunk 2]
```diff
@@ -153,4 +153,16 @@ impl<T: ComposedTransitionSystem> TransitionSystem for T {
...
+        let (left, right) = self.get_children();
+        let mut clocks = HashMap::new();
+        let mut ints = HashMap::new();
+        for decl in [left.get_combined_decls(), right.get_combined_decls()] {
+            clocks.extend(decl.clocks);
+            ints.extend(decl.ints)
+        }
+
+        Declarations { ints, clocks }
+    }
```
Comment (t-lohse) [inline src/TransitionSystems/common.rs:167]: Couldn't this just be a default implementation on `TransitionSystem`, and then only create a new for `CompilledComponent`?

File: src/System/extract_system_rep.rs [hunk 3]
```diff
@@ -149,6 +149,21 @@ impl SystemRecipe {
             },
         }
     }
+
+    /// Returns the components in [SystemRecipe] as a [Vec]<[Component]> in the same order as they visited.
+    pub fn get_components(&self) -> Vec<Component> {
```
Comment (t-lohse) [inline src/System/extract_system_rep.rs:154]: Never used?

File: src/System/extract_state.rs [hunk 4]
```diff
@@ -21,104 +18,81 @@ pub fn get_state(
...
-            Ok(State::create(locationtuple, zone))
         }
-        _ => panic!("Expected QueryExpression::State, but got {:?}", state_query),
+
+        Ok(State::create(
+            build_location_tuple(&mut locations.iter(), machine, system)?,
+            create_zone_given_constraints(clock.as_deref(), system)?,
+        ))
+    } else {
+        panic!("Expected QueryExpression::State, but got {:?}", state_query)
```
Comment (t-lohse) [inline src/System/extract_state.rs:36]: Why not just return `Err`, since the function returns `Result`?

Comment (Casper-NS) [reply src/System/extract_state.rs:36]: @Puvikaran2001 Henrik said to tag u

**Review by Casper-NS — COMMENTED** [review]
Not sure i completely understand it, but as far as i can tell it looks good. Tag me again when comments have been fixed and i will have a second look and hopefully approve.

File: src/TransitionSystems/common.rs (hunk 2 above)
Comment (Puvikaran2001) [reply src/TransitionSystems/common.rs]: Fixed @t-lohse

File: src/System/extract_system_rep.rs (hunk 3 above)
Comment (Puvikaran2001) [reply src/System/extract_system_rep.rs]: Fixed @t-lohse

File: src/System/extract_state.rs (hunk 4 above)
Comment (Puvikaran2001) [reply src/System/extract_state.rs]: Fixed @t-lohse

File: src/ModelObjects/component.rs [hunk 5]
```diff
@@ -75,6 +75,12 @@ impl Component {
             panic!("Unable to retrieve location based on id: {}", name)
         }
     }
+
+    /// Checks if a location exists in the component, by checking the given location `id` match any location in the component
+    pub fn location_exists(self, id: &str) -> bool {
+        self.get_locations().iter().any(|x| x.id.as_str() == id)
+    }
```
Comment (t-lohse) [inline src/ModelObjects/component.rs:82]: unused?

File: src/TransitionSystems/quotient.rs [hunk 6]
```diff
@@ -452,6 +453,21 @@ impl TransitionSystem for Quotient {
...
+            self.S.get_combined_decls(),
+            self.T.get_combined_decls(),
+            self.decls.clone(),
+        ] {
+            clocks.extend(decl.clocks);
+            ints.extend(decl.ints)
+        }
+
+        Declarations { ints, clocks }
+    }
```
Comment (t-lohse) [inline src/TransitionSystems/quotient.rs:470]: Doesn't this do the same as the default implementation?

File: src/ModelObjects/component.rs (hunk 5 above)
Comment (Puvikaran2001) [reply src/ModelObjects/component.rs]: This implementation is not used, so we have removed it now :) @t-lohse

File: src/TransitionSystems/quotient.rs (hunk 6 above)
Comment (Puvikaran2001) [reply src/TransitionSystems/quotient.rs]: Quotient has also a declaration, but I have changed the default implementation get_combined_decls, so this code is also removed :) @t-lohse

**Review by t-lohse — APPROVED** [review]

**Review by Casper-NS — APPROVED** [review]
Nice work, lgtm

**Pull request merged**